File: GameServer/Proto/LuaProto/testProto.lua
--新建协议
testProto = { ProtoCode = 17001, id = 0 }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
testProto.__index = testProto;

function testProto.New()
    local self = { }; --初始化self
    setmetatable(self, testProto); --将self的元表设定为Class
    return self;
end


--发送协议
function testProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteInt(proto.id);

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function testProto.GetProto(buffer)

    local proto = testProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.id = ms:ReadInt();

    ms:Dispose();
    return proto;
end

File: GameServer/Proto/CSharpProto/TestProto.cs

using System.Collections;

public struct TestProto : IProto
{
    public ushort ProtoCode { get { return 888; } }
    public int Id;
    public string Name;
    public int Type;
    public float price;


    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode); 
            ms.WriteInt(Id);
            ms.WriteUTF8String(Name);
            ms.WriteInt(Type);
            ms.WriteFloat(price);
            return ms.ToArray();
        }
    }


    public static TestProto GetProto(byte[] buffer)
    {
        TestProto pro = new TestProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            pro.Id = ms.ReadInt();
            pro.Name = ms.ReadUTF8String();
            pro.Type = ms.ReadInt();
            pro.price = ms.ReadFloat();

            return pro;
        }
    }
}
